File: lang/java/cxlib/src/test/java/cx/ConformanceTest.java
package cx;

import org.junit.jupiter.api.*;
import static org.junit.jupiter.api.Assertions.*;

import java.io.*;
import java.nio.file.*;
import java.util.*;

/**
 * CX Java conformance runner.
 * Parses conformance/*.txt and checks all test cases.
 */
public class ConformanceTest {

    // ── suite parser ───────────────────────────────────────────────────────────

    static class TestCase {
        String name;
        Map<String, String> sections = new LinkedHashMap<>();
        TestCase(String name) { this.name = name; }
    }

    static List<TestCase> parseSuite(Path path) throws IOException {
        List<TestCase> tests = new ArrayList<>();
        TestCase cur = null;
        String section = null;
        List<String> buf = new ArrayList<>();

        for (String raw : Files.readAllLines(path)) {
            String line = raw;
            if (line.startsWith("=== test:")) {
                flush(cur, section, buf);
                if (cur != null) tests.add(cur);
                cur = new TestCase(line.substring(9).trim());
                section = null;
            } else if (line.startsWith("--- ") && cur != null) {
                flush(cur, section, buf);
                section = line.substring(4).trim();
            } else if (section != null && cur != null) {
                buf.add(line);
            }
        }
        flush(cur, section, buf);
        if (cur != null) tests.add(cur);
        return tests;
    }

    static void flush(TestCase cur, String section, List<String> buf) {
        if (cur != null && section != null) {
            List<String> lines = new ArrayList<>(buf);
            while (!lines.isEmpty() && lines.get(0).trim().isEmpty())            lines.remove(0);
            while (!lines.isEmpty() && lines.get(lines.size()-1).trim().isEmpty()) lines.remove(lines.size()-1);
            cur.sections.put(section, String.join("\n", lines));
        }
        buf.clear();
    }

    // ── simple JSON value equality (structural, order-insensitive for objects) ─

    static Object parseJson(String s) {
        return new JsonParser(s.trim()).parse();
    }

    static boolean jsonEqual(Object a, Object b) {
        if (a == null && b == null) return true;
        if (a == null || b == null) return false;
        if (!a.getClass().equals(b.getClass())) {
            // allow int/long comparisons
            if ((a instanceof Number) && (b instanceof Number))
                return ((Number)a).doubleValue() == ((Number)b).doubleValue();
            return false;
        }
        if (a instanceof Map) {
            @SuppressWarnings("unchecked") Map<String,Object> ma = (Map<String,Object>)a;
            @SuppressWarnings("unchecked") Map<String,Object> mb = (Map<String,Object>)b;
            if (!ma.keySet().equals(mb.keySet())) return false;
            for (String k : ma.keySet()) if (!jsonEqual(ma.get(k), mb.get(k))) return false;
            return true;
        }
        if (a instanceof List) {
            @SuppressWarnings("unchecked") List<Object> la = (List<Object>)a;
            @SuppressWarnings("unchecked") List<Object> lb = (List<Object>)b;
            if (la.size() != lb.size()) return false;
            for (int i = 0; i < la.size(); i++) if (!jsonEqual(la.get(i), lb.get(i))) return false;
            return true;
        }
        return a.equals(b);
    }

    // tiny recursive-descent JSON parser
    static class JsonParser {
        final String s;
        int pos;
        JsonParser(String s) { this.s = s; }

        Object parse() {
            skipWs();
            if (pos >= s.length()) return null;
            char c = s.charAt(pos);
            if (c == '{') return parseObject();
            if (c == '[') return parseArray();
            if (c == '"') return parseString();
            if (c == 't') { pos += 4; return Boolean.TRUE; }
            if (c == 'f') { pos += 5; return Boolean.FALSE; }
            if (c == 'n') { pos += 4; return null; }
            return parseNumber();
        }

        void skipWs() { while (pos < s.length() && s.charAt(pos) <= ' ') pos++; }

        Map<String,Object> parseObject() {
            Map<String,Object> m = new LinkedHashMap<>();
            pos++; // {
            skipWs();
            while (pos < s.length() && s.charAt(pos) != '}') {
                String key = parseString();
                skipWs(); pos++; // :
                Object val = parse();
                m.put(key, val);
                skipWs();
                if (pos < s.length() && s.charAt(pos) == ',') { pos++; skipWs(); }
            }
            if (pos < s.length()) pos++; // }
            return m;
        }

        List<Object> parseArray() {
            List<Object> a = new ArrayList<>();
            pos++; // [
            skipWs();
            while (pos < s.length() && s.charAt(pos) != ']') {
                a.add(parse());
                skipWs();
                if (pos < s.length() && s.charAt(pos) == ',') { pos++; skipWs(); }
            }
            if (pos < s.length()) pos++; // ]
            return a;
        }

        String parseString() {
            pos++; // "
            StringBuilder sb = new StringBuilder();
            while (pos < s.length()) {
                char c = s.charAt(pos++);
                if (c == '"') break;
                if (c == '\\') {
                    c = s.charAt(pos++);
                    switch (c) {
                        case '"': sb.append('"'); break;
                        case '\\': sb.append('\\'); break;
                        case '/':  sb.append('/');  break;
                        case 'n':  sb.append('\n'); break;
                        case 'r':  sb.append('\r'); break;
                        case 't':  sb.append('\t'); break;
                        case 'u':
                            int cp = Integer.parseInt(s.substring(pos, pos+4), 16);
                            sb.appendCodePoint(cp); pos += 4; break;
                        default: sb.append(c);
                    }
                } else { sb.append(c); }
            }
            return sb.toString();
        }

        Number parseNumber() {
            int start = pos;
            if (pos < s.length() && s.charAt(pos) == '-') pos++;
            while (pos < s.length() && Character.isDigit(s.charAt(pos))) pos++;
            boolean isFloat = false;
            if (pos < s.length() && s.charAt(pos) == '.') { isFloat = true; pos++; while (pos < s.length() && Character.isDigit(s.charAt(pos))) pos++; }
            if (pos < s.length() && (s.charAt(pos) == 'e' || s.charAt(pos) == 'E')) {
                isFloat = true; pos++;
                if (pos < s.length() && (s.charAt(pos) == '+' || s.charAt(pos) == '-')) pos++;
                while (pos < s.length() && Character.isDigit(s.charAt(pos))) pos++;
            }
            String num = s.substring(start, pos);
            if (isFloat) return Double.parseDouble(num);
            try { return Long.parseLong(num); } catch (NumberFormatException e) { return Double.parseDouble(num); }
        }
    }

    // ── test runner ────────────────────────────────────────────────────────────

    static String dispatch(String inFmt, String outFmt, String input) {
        switch (inFmt + ":" + outFmt) {
            case "cx:cx":     return CxLib.toCx(input);
            case "cx:xml":    return CxLib.toXml(input);
            case "cx:ast":    return CxLib.toAst(input);
            case "cx:json":   return CxLib.toJson(input);
            case "cx:yaml":   return CxLib.toYaml(input);
            case "cx:toml":   return CxLib.toToml(input);
            case "cx:md":     return CxLib.toMd(input);
            case "xml:cx":    return CxLib.xmlToCx(input);
            case "xml:xml":   return CxLib.xmlToXml(input);
            case "xml:ast":   return CxLib.xmlToAst(input);
            case "xml:json":  return CxLib.xmlToJson(input);
            case "xml:yaml":  return CxLib.xmlToYaml(input);
            case "xml:toml":  return CxLib.xmlToToml(input);
            case "xml:md":    return CxLib.xmlToMd(input);
            case "json:cx":   return CxLib.jsonToCx(input);
            case "json:xml":  return CxLib.jsonToXml(input);
            case "json:ast":  return CxLib.jsonToAst(input);
            case "json:json": return CxLib.jsonToJson(input);
            case "json:yaml": return CxLib.jsonToYaml(input);
            case "json:toml": return CxLib.jsonToToml(input);
            case "json:md":   return CxLib.jsonToMd(input);
            case "yaml:cx":   return CxLib.yamlToCx(input);
            case "yaml:xml":  return CxLib.yamlToXml(input);
            case "yaml:ast":  return CxLib.yamlToAst(input);
            case "yaml:json": return CxLib.yamlToJson(input);
            case "yaml:yaml": return CxLib.yamlToYaml(input);
            case "yaml:toml": return CxLib.yamlToToml(input);
            case "yaml:md":   return CxLib.yamlToMd(input);
            case "toml:cx":   return CxLib.tomlToCx(input);
            case "toml:xml":  return CxLib.tomlToXml(input);
            case "toml:ast":  return CxLib.tomlToAst(input);
            case "toml:json": return CxLib.tomlToJson(input);
            case "toml:yaml": return CxLib.tomlToYaml(input);
            case "toml:toml": return CxLib.tomlToToml(input);
            case "toml:md":   return CxLib.tomlToMd(input);
            case "md:cx":     return CxLib.mdToCx(input);
            case "md:xml":    return CxLib.mdToXml(input);
            case "md:ast":    return CxLib.mdToAst(input);
            case "md:json":   return CxLib.mdToJson(input);
            case "md:yaml":   return CxLib.mdToYaml(input);
            case "md:toml":   return CxLib.mdToToml(input);
            case "md:md":     return CxLib.mdToMd(input);
            default: throw new IllegalArgumentException("no dispatch for " + inFmt + ":" + outFmt);
        }
    }

    static List<String> runTest(TestCase t) {
        List<String> failures = new ArrayList<>();
        Map<String, String> s = t.sections;

        String src = null, inFmt = null;
        for (String[] pair : new String[][]{
                {"in_cx","cx"}, {"in_xml","xml"}, {"in_json","json"},
                {"in_yaml","yaml"}, {"in_toml","toml"}, {"in_md","md"}}) {
            if (s.containsKey(pair[0])) { src = s.get(pair[0]); inFmt = pair[1]; break; }
        }
        if (inFmt == null) return failures;

        final String finalSrc = src, finalInFmt = inFmt;

        // helper
        String[] call = new String[2]; // [out, err]
        java.util.function.Consumer<String> doCall = outFmt -> {
            try   { call[0] = dispatch(finalInFmt, outFmt, finalSrc); call[1] = null; }
            catch (Exception e) { call[0] = null; call[1] = e.getMessage(); }
        };

        if (s.containsKey("out_ast")) {
            doCall.accept("ast");
            if (call[1] != null) { failures.add("out_ast parse error: " + call[1]); }
            else {
                Object exp = parseJson(s.get("out_ast")), got = parseJson(call[0]);
                if (!jsonEqual(exp, got)) failures.add("out_ast mismatch\n  expected: " + s.get("out_ast") + "\n  got:      " + call[0]);
            }
        }
        if (s.containsKey("out_xml")) {
            doCall.accept("xml");
            if (call[1] != null) failures.add("out_xml parse error: " + call[1]);
            else if (!s.get("out_xml").trim().equals(call[0].trim())) failures.add("out_xml mismatch\n  expected:\n" + s.get("out_xml") + "\n  got:\n" + call[0]);
        }
        if (s.containsKey("out_cx")) {
            doCall.accept("cx");
            if (call[1] != null) failures.add("out_cx parse error: " + call[1]);
            else if (!s.get("out_cx").trim().equals(call[0].trim())) failures.add("out_cx mismatch\n  expected:\n" + s.get("out_cx") + "\n  got:\n" + call[0]);
        }
        if (s.containsKey("out_json")) {
            doCall.accept("json");
            if (call[1] != null) { failures.add("out_json parse error: " + call[1]); }
            else {
                Object exp = parseJson(s.get("out_json")), got = parseJson(call[0]);
                if (!jsonEqual(exp, got)) failures.add("out_json mismatch\n  expected: " + s.get("out_json") + "\n  got:      " + call[0]);
            }
        }
        if (s.containsKey("out_md")) {
            doCall.accept("md");
            if (call[1] != null) failures.add("out_md parse error: " + call[1]);
            else if (!s.get("out_md").trim().equals(call[0].trim())) failures.add("out_md mismatch\n  expected:\n" + s.get("out_md") + "\n  got:\n" + call[0]);
        }
        return failures;
    }

    static int runSuite(Path path) throws IOException {
        List<TestCase> tests = parseSuite(path);
        int passed = 0, failed = 0;
        for (TestCase t : tests) {
            List<String> failures;
            try   { failures = runTest(t); }
            catch (Exception e) { failures = List.of("runner exception: " + e.getMessage()); }
            if (failures.isEmpty()) { passed++; }
            else {
                failed++;
                System.out.println("FAIL  " + t.name);
                for (String f : failures)
                    for (String line : f.split("\n")) System.out.println("      " + line);
            }
        }
        System.out.println(path + ": " + passed + " passed, " + failed + " failed");
        return failed;
    }

    // ── JUnit 5 test methods ───────────────────────────────────────────────────

    static Path conformanceDir() {
        // tests run from lang/java/cxlib; conformance/ is at ../../../conformance/
        return Paths.get("../../../conformance");
    }

    @Test
    void testCore() throws IOException {
        int f = runSuite(conformanceDir().resolve("core.txt"));
        assertEquals(0, f, "core.txt had failures");
    }

    @Test
    void testExtended() throws IOException {
        int f = runSuite(conformanceDir().resolve("extended.txt"));
        assertEquals(0, f, "extended.txt had failures");
    }

    @Test
    void testXml() throws IOException {
        int f = runSuite(conformanceDir().resolve("xml.txt"));
        assertEquals(0, f, "xml.txt had failures");
    }

    @Test
    void testMd() throws IOException {
        int f = runSuite(conformanceDir().resolve("md.txt"));
        assertEquals(0, f, "md.txt had failures");
    }
}

File: lang/kotlin/cxlib/src/test/kotlin/cx/ConformanceTest.kt
package cx

import org.junit.jupiter.api.Test
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import kotlin.test.assertEquals

/**
 * CX Kotlin conformance runner.
 */
class ConformanceTest {

    // ── suite parser ───────────────────────────────────────────────────────────

    data class TestCase(val name: String, val sections: MutableMap<String, String> = mutableMapOf())

    fun parseSuite(path: Path): List<TestCase> {
        val tests = mutableListOf<TestCase>()
        var cur: TestCase? = null
        var section: String? = null
        val buf = mutableListOf<String>()

        fun flush() {
            val c = cur ?: return
            val s = section ?: return
            val lines = buf.toMutableList()
            while (lines.isNotEmpty() && lines.first().isBlank()) lines.removeFirst()
            while (lines.isNotEmpty() && lines.last().isBlank()) lines.removeLast()
            c.sections[s] = lines.joinToString("\n")
            buf.clear()
        }

        for (line in Files.readAllLines(path)) {
            when {
                line.startsWith("=== test:") -> {
                    flush()
                    cur?.let { tests.add(it) }
                    cur = TestCase(line.substring(9).trim())
                    section = null
                }
                line.startsWith("--- ") && cur != null -> {
                    flush()
                    section = line.substring(4).trim()
                }
                section != null && cur != null -> buf.add(line)
            }
        }
        flush()
        cur?.let { tests.add(it) }
        return tests
    }

    // ── tiny JSON parser + deep equality ──────────────────────────────────────

    private fun parseJson(s: String): Any? = JsonParser(s.trim()).parse()

    private fun jsonEqual(a: Any?, b: Any?): Boolean {
        if (a === b) return true
        if (a == null || b == null) return false
        if (a is Map<*,*> && b is Map<*,*>) {
            if (a.keys != b.keys) return false
            return a.keys.all { jsonEqual(a[it], b[it]) }
        }
        if (a is List<*> && b is List<*>) {
            if (a.size != b.size) return false
            return a.indices.all { jsonEqual(a[it], b[it]) }
        }
        if (a is Number && b is Number) return a.toDouble() == b.toDouble()
        return a == b
    }

    inner class JsonParser(val s: String) {
        var pos = 0
        fun parse(): Any? {
            skipWs()
            if (pos >= s.length) return null
            return when (s[pos]) {
                '{' -> parseObject()
                '[' -> parseArray()
                '"' -> parseString()
                't' -> { pos += 4; true }
                'f' -> { pos += 5; false }
                'n' -> { pos += 4; null }
                else -> parseNumber()
            }
        }
        fun skipWs() { while (pos < s.length && s[pos] <= ' ') pos++ }
        fun parseObject(): Map<String, Any?> {
            pos++; skipWs()
            val m = mutableMapOf<String, Any?>()
            while (pos < s.length && s[pos] != '}') {
                val k = parseString(); skipWs(); pos++ // :
                m[k] = parse(); skipWs()
                if (pos < s.length && s[pos] == ',') { pos++; skipWs() }
            }
            if (pos < s.length) pos++; return m
        }
        fun parseArray(): List<Any?> {
            pos++; skipWs()
            val a = mutableListOf<Any?>()
            while (pos < s.length && s[pos] != ']') {
                a.add(parse()); skipWs()
                if (pos < s.length && s[pos] == ',') { pos++; skipWs() }
            }
            if (pos < s.length) pos++; return a
        }
        fun parseString(): String {
            pos++
            val sb = StringBuilder()
            while (pos < s.length) {
                val c = s[pos++]
                if (c == '"') break
                if (c == '\\') {
                    when (val e = s[pos++]) {
                        '"'  -> sb.append('"')
                        '\\' -> sb.append('\\')
                        '/'  -> sb.append('/')
                        'n'  -> sb.append('\n')
                        'r'  -> sb.append('\r')
                        't'  -> sb.append('\t')
                        'u'  -> { sb.appendCodePoint(s.substring(pos, pos+4).toInt(16)); pos += 4 }
                        else -> sb.append(e)
                    }
                } else sb.append(c)
            }
            return sb.toString()
        }
        fun parseNumber(): Number {
            val start = pos
            if (pos < s.length && s[pos] == '-') pos++
            while (pos < s.length && s[pos].isDigit()) pos++
            var isFloat = false
            if (pos < s.length && s[pos] == '.') { isFloat = true; pos++; while (pos < s.length && s[pos].isDigit()) pos++ }
            if (pos < s.length && (s[pos] == 'e' || s[pos] == 'E')) {
                isFloat = true; pos++
                if (pos < s.length && (s[pos] == '+' || s[pos] == '-')) pos++
                while (pos < s.length && s[pos].isDigit()) pos++
            }
            val num = s.substring(start, pos)
            return if (isFloat) num.toDouble() else num.toLong()
        }
    }

    // ── dispatch ───────────────────────────────────────────────────────────────

    fun dispatch(inFmt: String, outFmt: String, input: String): String = when ("$inFmt:$outFmt") {
        "cx:cx"     -> CxLib.toCx(input);     "cx:xml"    -> CxLib.toXml(input)
        "cx:ast"    -> CxLib.toAst(input);    "cx:json"   -> CxLib.toJson(input)
        "cx:yaml"   -> CxLib.toYaml(input);   "cx:toml"   -> CxLib.toToml(input)
        "cx:md"     -> CxLib.toMd(input)
        "xml:cx"    -> CxLib.xmlToCx(input);  "xml:xml"   -> CxLib.xmlToXml(input)
        "xml:ast"   -> CxLib.xmlToAst(input); "xml:json"  -> CxLib.xmlToJson(input)
        "xml:yaml"  -> CxLib.xmlToYaml(input);"xml:toml"  -> CxLib.xmlToToml(input)
        "xml:md"    -> CxLib.xmlToMd(input)
        "json:cx"   -> CxLib.jsonToCx(input); "json:xml"  -> CxLib.jsonToXml(input)
        "json:ast"  -> CxLib.jsonToAst(input);"json:json" -> CxLib.jsonToJson(input)
        "json:yaml" -> CxLib.jsonToYaml(input);"json:toml"-> CxLib.jsonToToml(input)
        "json:md"   -> CxLib.jsonToMd(input)
        "yaml:cx"   -> CxLib.yamlToCx(input); "yaml:xml"  -> CxLib.yamlToXml(input)
        "yaml:ast"  -> CxLib.yamlToAst(input);"yaml:json" -> CxLib.yamlToJson(input)
        "yaml:yaml" -> CxLib.yamlToYaml(input);"yaml:toml"-> CxLib.yamlToToml(input)
        "yaml:md"   -> CxLib.yamlToMd(input)
        "toml:cx"   -> CxLib.tomlToCx(input); "toml:xml"  -> CxLib.tomlToXml(input)
        "toml:ast"  -> CxLib.tomlToAst(input);"toml:json" -> CxLib.tomlToJson(input)
        "toml:yaml" -> CxLib.tomlToYaml(input);"toml:toml"-> CxLib.tomlToToml(input)
        "toml:md"   -> CxLib.tomlToMd(input)
        "md:cx"     -> CxLib.mdToCx(input);   "md:xml"    -> CxLib.mdToXml(input)
        "md:ast"    -> CxLib.mdToAst(input);  "md:json"   -> CxLib.mdToJson(input)
        "md:yaml"   -> CxLib.mdToYaml(input); "md:toml"   -> CxLib.mdToToml(input)
        "md:md"     -> CxLib.mdToMd(input)
        else -> throw IllegalArgumentException("no dispatch for $inFmt:$outFmt")
    }

    // ── test runner ────────────────────────────────────────────────────────────

    fun runTest(t: TestCase): List<String> {
        val failures = mutableListOf<String>()
        val s = t.sections

        val (src, inFmt) = listOf(
            "in_cx" to "cx", "in_xml" to "xml", "in_json" to "json",
            "in_yaml" to "yaml", "in_toml" to "toml", "in_md" to "md"
        ).firstOrNull { (k, _) -> k in s }
            ?.let { (k, fmt) -> s[k]!! to fmt } ?: return failures

        fun call(outFmt: String): Pair<String?, String?> = try {
            dispatch(inFmt, outFmt, src) to null
        } catch (e: Exception) { null to e.message }

        if ("out_ast" in s) {
            val (out, err) = call("ast")
            if (err != null) failures.add("out_ast parse error: $err")
            else if (!jsonEqual(parseJson(s["out_ast"]!!), parseJson(out!!)))
                failures.add("out_ast mismatch\n  expected: ${s["out_ast"]}\n  got:      $out")
        }
        if ("out_xml" in s) {
            val (out, err) = call("xml")
            if (err != null) failures.add("out_xml parse error: $err")
            else if (s["out_xml"]!!.trim() != out!!.trim())
                failures.add("out_xml mismatch\n  expected:\n${s["out_xml"]}\n  got:\n$out")
        }
        if ("out_cx" in s) {
            val (out, err) = call("cx")
            if (err != null) failures.add("out_cx parse error: $err")
            else if (s["out_cx"]!!.trim() != out!!.trim())
                failures.add("out_cx mismatch\n  expected:\n${s["out_cx"]}\n  got:\n$out")
        }
        if ("out_json" in s) {
            val (out, err) = call("json")
            if (err != null) failures.add("out_json parse error: $err")
            else if (!jsonEqual(parseJson(s["out_json"]!!), parseJson(out!!)))
                failures.add("out_json mismatch\n  expected: ${s["out_json"]}\n  got:      $out")
        }
        if ("out_md" in s) {
            val (out, err) = call("md")
            if (err != null) failures.add("out_md parse error: $err")
            else if (s["out_md"]!!.trim() != out!!.trim())
                failures.add("out_md mismatch\n  expected:\n${s["out_md"]}\n  got:\n$out")
        }
        return failures
    }

    fun runSuite(path: Path): Int {
        val tests = parseSuite(path)
        var passed = 0; var failed = 0
        for (t in tests) {
            val failures = try { runTest(t) } catch (e: Exception) { listOf("runner exception: ${e.message}") }
            if (failures.isEmpty()) passed++
            else {
                failed++
                println("FAIL  ${t.name}")
                failures.forEach { f -> f.split("\n").forEach { println("      $it") } }
            }
        }
        println("$path: $passed passed, $failed failed")
        return failed
    }

    private fun conformanceDir(): Path = Paths.get("../../../conformance")

    @Test fun testCore()     { assertEquals(0, runSuite(conformanceDir().resolve("core.txt")), "core.txt") }
    @Test fun testExtended() { assertEquals(0, runSuite(conformanceDir().resolve("extended.txt")), "extended.txt") }
    @Test fun testXml()      { assertEquals(0, runSuite(conformanceDir().resolve("xml.txt")), "xml.txt") }
    @Test fun testMd()       { assertEquals(0, runSuite(conformanceDir().resolve("md.txt")), "md.txt") }
}
